coriolis/providers/backup_writers.py

```diff
@@ -628,29 +711,124 @@ def close(self):
-            )
-        self._crt_dir = cert_dir
+                "Either ssh or writer_creds must be specified")
+        if self._ssh and self._certificates:
+            raise exception.CoriolisException(
+                "ssh and writer_creds are mutually exclusive")
+        self._cert_paths = None
+
+    @classmethod
+    def from_connection_info(cls, conn_info, volumes_info):
+        """Instantiate a HTTP backup writer from connection info.
+
+        Connection info has the following schema:
+
+        {
+            # IP address or hostname where we can reach the backup writer
+            "ip": "192.168.0.1",
+            # Backup writer port
+            "port": 4433,
+
+            # ssh_conn_info is only needed in environments where the HTTP
+            # backup writer cannot be set up via userdata. This can be
+            # used to copy the coriolis-writer binary, create certificates
+            # and initialize everything. This option and the "certificates"
+            # option are mutually exclusive. Set to None if the writer has
+            # already been set up via userdata, or other means.
+            "ssh_conn_info": {
+                # The IP here is usually the same as the IP/hostname of the
+                # backup writer, shown above.
+                "ip": "192.168.0.1",
+                # SSH port. Defaults to 22.
+                "port": 22,
+                "username": "example",
+                # Password is only needed if no pkey is specified
+                "password": "super secret password",
+                # pkey can be omited if password is used. Pkey must be a
+                # PEM formatted OpenSSH private key, paramiko.RSA private key
+                # or None
+                "pkey": "RSA_PRIVATE_KEY"
+            },
+
+            # The "certificates" field and the "ssh_conn_info" are mutually
+            # exclusive. Either the writer is set up via userdata, or we set
+            # it up ourselves using a SSH connection. But whether or not it's
+            # set up must be unambiguous. Setting this option means we already
+            # have the certificates, and the coriolis-writer binary is already
+            # set up and started
+            "certificates": {
+                # PEM encoded client certificate
+                "client_crt": "",
+                # PEM encoded client private key
+                "client_key": "",
+                # PEM encoded CA certificate we use to validate the server
+                "ca_crt": ""
+            }
+        }
+        """
+        ip = conn_info.get("ip")
+        port = conn_info.get("port")
+
+        required = ["ip", "port"]
+        if not all([ip, port]):
+            raise exception.CoriolisException(
+                "Missing required connection info: %s" % ", ".join(required))
+
+        ssh_cli = None
+        ssh = conn_info.get("ssh_conn_info")
+        certs = conn_info.get("certificates")
```

Comment:
I was not saying we should move this to the constructor, I was saying that we should check if both `certs` and `ssh ` was provided and error out here, instead of actually opening the SSH connection here:
https://github.com/cloudbase/coriolis/pull/112/files#diff-ae50c984218b903f95f34c20c8660dbeR818

only to have this exact thing checked and fail here:
https://github.com/cloudbase/coriolis/pull/112/files#diff-ae50c984218b903f95f34c20c8660dbeR731

Reply:
You are correct. This whole mutually exclusive thing bugs me to be honest, and needs to go. I will split up the ```HTTP``` backup writer service bootstrap to its own class, which will return the proper connection info. That in turn will be used to instantiate the ```HTTPBackupWriter()```.

This way providers that want to use the ```HTTPBackupWriter()``` can choose to either bootstrap the thing using ```userdata``` or call the soon to be written ```HTTPBackupWriterBootstrapper()``` and get proper connection info, and avoid this whole mess.

Reply:
Okay. Done. Should be a bit more sane. Added benefit is, any provider can now easily bootstrap the ```HTTPBackupWriter()``` by simply calling:

```python
writer_conn_info = HTTPBackupWriterBootstrapper(
    ssh_conn_info, writer_port).setup_writer()
```